templates/verifier_groth16.vy.ejs

```diff
@@ -0,0 +1,189 @@
+          8495653923123431417604973247489272438418190587263600148770280649306958101930
+        ]}
+      )
+
+
+# @dev return r the negation of p, i.e. p.addition(p.negate()) should be zero.
+@internal
+@view
+def negate(p: G1Point) -> G1Point:
+    if (p.X == 0) and (p.Y == 0):
+        return G1Point({X: 0, Y: 0})
+
+    # Validate input or revert
+    assert (p.X < BASE_MODULUS) or (p.Y < BASE_MODULUS), "invalid proof"
+
+    # We know p.Y > 0 and p.Y < BASE_MODULUS.
+    return G1Point({X: p.X, Y: BASE_MODULUS - p.Y})
+
+
+# @dev return r the sum of two points of G1@return r the sum of two points of G1
+@internal
+@view
+def addition(p1: G1Point, p2: G1Point) -> G1Point:
+    # By EIP-196 all input is validated to be less than the BASE_MODULUS and form points
+    # on the curve.
+    a: uint256[2] = [p1.X, p1.Y]
+    b: uint256[2] = [p2.X, p2.Y]
+    result : uint256[2] = ecadd(a, b)
+    r : G1Point = G1Point({X: result[0], Y: result[1]})
+    return r
+
+
+# @dev return r the product of a point on G1 and a scalar, i.e.
+#      p == p.scalar_mul(1) and p.addition(p) == p.scalar_mul(2) for all points p.
+@internal
+@view
+def scalar_mul(p: G1Point, s: uint256) -> G1Point:
+    a: uint256[2] = [p.X, p.Y]
+    result: uint256[2] = ecmul(a, s)
+    r: G1Point = G1Point({X: result[0], Y: result[1]})
+    return r
+
+
+# @dev Asserts the pairing check
+#      e(p1[0], p2[0]) *  .... * e(p1[n], p2[n]) == 1
+#      For example pairing([P1(), P1().negate()], [P2(), P2()]) should succeed
+#      this function supports n=N_INPUTS maximum
+@internal
+@view
+def pairing_check_4(p1 : DynArray[G1Point, 4], p2: DynArray[G2Point, 4]) -> bool:
+    # By EIP-197 all input is verified to be less than the BASE_MODULUS and form elements in their
+    # respective groups of the right order.
+
+    assert len(p1) == len(p2), "invalid pairing"
+
+    input: DynArray[bytes32, 24] = []
+    for i in range(4): 
+        input.append(convert(p1[i].X, bytes32))
+        input.append(convert(p1[i].Y, bytes32))
+        input.append(convert(p2[i].X[0], bytes32))
+        input.append(convert(p2[i].X[1], bytes32))
+        input.append(convert(p2[i].Y[0], bytes32))
+        input.append(convert(p2[i].Y[1], bytes32))
+
+    converted: Bytes[768] = concat(input[0], input[1], input[2], input[3], input[4],input[5], 
+                                   input[6], input[7], input[8], input[9], input[10], input[11],
+                                   input[12], input[13], input[14], input[15], input[16], input[17], 
+                                   input[18],input[19], input[20], input[21], input[22], input[23])
+
+    response: Bytes[32] = b''
+
+    # raw_call to ecpairing contract
+    response = raw_call(0x0000000000000000000000000000000000000008, 
+                        converted, 
+                        max_outsize=32, 
+                        is_static_call=True
+                        )
+
+    return _abi_decode(response, bool)
```

Comment:
maybe a comment explaining that ecpairing returns 0x00 or 0x01 depending on success would be good here.